carriers/dhl.js

```diff
@@ -0,0 +1,127 @@
+const async = require('async');
+const request = require('request');
+const geography = require('../util/geography');
+const moment = require('moment-timezone');
+const checkDigit = require('../util/checkDigit');
+
+function DHL(options) {
+    this.isTrackingNumberValid = function(trackingNumber) {
+        // remove whitespace
+        trackingNumber = trackingNumber.replace(/\s/g, '');
+        trackingNumber = trackingNumber.toUpperCase();
+
+        if ([/^93612\d{17}$/, /^92612\d{17}$/, /^94748\d{17}$/, /^93748\d{17}$/, /^92748\d{17}$/].some(regex => regex.test(trackingNumber))) {
+            return checkDigit(trackingNumber, [3, 1], 10);
+        }
+        if (/^420\d{27}$/.test(trackingNumber)) {
+            return checkDigit(trackingNumber.match(/^420\d{5}(\d{22})$/)[1], [3, 1], 10);
+        }
+        if (/^420\d{31}$/.test(trackingNumber)) {
+            if (checkDigit(trackingNumber.match(/^420\d{9}(\d{22})$/)[1], [3, 1], 10)) {
+                return true;
+            } else if (checkDigit(trackingNumber.match(/^420\d{5}(\d{26})$/)[1], [3, 1], 10)) {
+                return true;
+            }
+        }
+
+        return false;
+    };
+
+    this.track = function(trackingNumber, callback) {
+        const req = {
+            url: `https://api-eu.dhl.com/track/shipments?trackingNumber=${trackingNumber}`,
+            method: 'GET',
+            headers: {
+                'DHL-API-Key': `${options.DHL_API_Key}`
+            },
+            timeout: 5000
+        };
+
+        async.retry(function(callback) {
+            request(req, callback);
+        }, function(err, res) {
+            const response = JSON.parse(res.body);
+
+            if (err) {
+                return callback(err);
+                //invalid credentials or invalid tracking number
+            } else if (response.status === 401 || response.status === 404) {
+                return callback(new Error (response.detail));
+            }
+
+            const results = {
+                events: []
+            };
+
+            const scanDetails = JSON.parse(res.body).shipments[0].events;
+
+            scanDetails.forEach(scanDetail => {
+                scanDetail.address = {
+                    city: scanDetail.location != undefined ? scanDetail.location.address.addressLocality : '',
+                    zip: scanDetail.location != undefined ? scanDetail.location.address.postalCode : ''
+                }
+                scanDetail.location = geography.addressToString(scanDetail.address);
+            });
+
+            // Get unqiue array of locations
+            const locations = Array.from(new Set(scanDetails.map(scanDetail => scanDetail.location)));
+
+            // Lookup each location
+            async.mapLimit(locations, 10, function (location, callback) {
```

Comment:
`function(` instead of `function (` throughout.
70, 71, 80, many in the test file